File: c++/leetcode/50. Pow(x, n).cpp
class Solution
{
public:
  double myPow(double x, int n)
  {
    if (n == 0)
      return 1;

    double partial = myPow(x, n / 2);

    if (n % 2 == 0)
      return partial * partial;
    else
    {
      if (n < 0)
      {
        return (partial * partial) / x;
      }

      return x * partial * partial;
    }
  }
};

File: python/leetcode/50. Pow(x, n).py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1

        partial = self.myPow(x, int(n / 2))

        if n % 2 == 0:
            return partial * partial
        else:
            if n < 0:
                return (partial * partial) / x
            return x * partial * partial
